app/validators/post_category_validator.rb

```diff
@@ -0,0 +1,7 @@
+class PostCategoryValidator < ActiveModel::Validator
+  def validate(record)
+    if record.category.has_children?
+      record.errors[:category_id] << "子カテゴリを持つカテゴリ「#{record.category.full_name}」に記事を登録することはできません。"
```

Comment:
In this project we don't use i18n for all messages.
If we don't need Japanese message here, I will translate to English.